### PR TITLE
[SRT] Add dynamic cuDNN SDPA for vision attention

### PR DESCRIPTION
**Pull request by mickqian**
## Summary

Add two multimodal vision attention backends for SRT:

- `cudnn_sdpa`: explicitly runs PyTorch SDPA with `SDPBackend.CUDNN_ATTENTION`.
- `dynamic_cudnn_sdpa`: chooses cuDNN SDPA only for dense equal-length no-mask ViT attention shapes where it benchmarks faster, and falls back to the platform default backend otherwise.

On Hopper, the default multimodal vision backend changes from `fa3` to `dynamic_cudnn_sdpa`. Explicit `--mm-attention-backend fa3` still forces the previous behavior.

The dynamic policy is intentionally conservative:

- Use cuDNN SDPA only for CUDA Hopper, fp16/bf16, equal Q/K heads, no `cu_seqlens`, no attention mask, no window/sink attention, `head_dim in {64, 80, 128}`, and `seq_len <= 1025`.
- Fall back to the previous platform default for varlen, masked, windowed, sink, longer-sequence, or non-Hopper paths.

## Benchmark

H100 80GB, CUDA 13.0, PyTorch `2.11.0+cu130`, bf16 random QKV, `heads=16`, `head_dim=80`, median CUDA event latency.

Dense no-mask ViT attention:

| shape | fa3 | dynamic_cudnn_sdpa | change |
| --- | ---: | ---: | ---: |
| bs=1, seq=576 | 0.0667 ms | 0.0641 ms | +4.1% |
| bs=4, seq=1024 | 0.1164 ms | 0.0987 ms | +18.0% |
| bs=4, seq=2304 | 0.2686 ms | 0.2733 ms | fallback to FA3 |

Varlen/masked control with `cu_seqlens` passed:

| shape | fa3 | dynamic_cudnn_sdpa | output diff |
| --- | ---: | ---: | ---: |
| bs=1, seq=576 | 0.0656 ms | 0.0674 ms | 0 |
| bs=4, seq=1024 | 0.1173 ms | 0.1193 ms | 0 |
| bs=4, seq=2304 | 0.2678 ms | 0.2782 ms | 0 |

The control verifies dynamic fallback avoids the slower dense-mask cuDNN path.

Real E2E, `OpenGVLab/InternVL2_5-2B` with dummy weights on H100, 64 image prompts, 448x448, input_len=64, output_len=16, max_concurrency=4:

| backend | req/s | output tok/s | mean e2e latency | mean TTFT | change vs fa3 |
| --- | ---: | ---: | ---: | ---: | ---: |
| fa3 | 13.72 | 121.10 | 287.55 ms | 170.07 ms | baseline |
| cudnn_sdpa | 13.90 | 122.73 | 283.69 ms | 168.28 ms | +1.35% req/s, -1.34% latency |
| dynamic_cudnn_sdpa | 14.27 | 126.00 | 275.17 ms | 152.00 ms | +4.05% req/s, -4.31% latency |

The real InternVL path passes `SingletonCache` for dense full-image attention; this PR now treats that as the dense no-mask path so dynamic dispatch can select cuDNN SDPA. The 448x448 InternVL shape is 1025 tokens including the class token, so the conservative cutoff is `seq_len <= 1025`.

## Validation

- `pre-commit run --files python/sglang/srt/layers/attention/vision.py python/sglang/srt/server_args.py`
- H100 microbenchmarks above
- H100 InternVL2_5-2B real E2E benchmark above

## Follow-up validation after merging main

Tested PR HEAD `be04e1c581d5ea4f94146af5d7cf01b4e0eba6bb` on an H100 80GB with CUDA 13.0, PyTorch `2.11.0+cu130`, cuDNN 9.19, and BF16.

### Dispatch finding

The original InternVL E2E result is no longer representative after merging main:

- InternVL now materializes a tensor `cu_seqlens` and prepares `max_seqlen` once before the encoder layer loop.
- `VisionDynamicCudnnSdpaAttention` treats any tensor `cu_seqlens` as varlen and falls back to FA3.
- Therefore, the current `dynamic_cudnn_sdpa` path does not select cuDNN for the InternVL workload used in the original PR description.

The precomputed `max_seqlen` also removes the per-layer GPU-to-host `.item()` synchronization that was included in the earlier FA3 measurements. For example, at `bs=1, seq=576, heads=16, head_dim=80`, FA3 takes `0.0681 ms` when recomputing max seqlen, but only `0.0420 ms` with the current precomputed metadata. Direct cuDNN SDPA takes `0.0572 ms`.

### Expanded attention microbenchmark

Median CUDA event latency:

| shape `(bs, seq, heads, head_dim)` | current FA3 | cuDNN SDPA | cuDNN vs FA3 |
| --- | ---: | ---: | ---: |
| `(1, 576, 16, 80)` | 0.0420 ms | 0.0572 ms | -26.5% |
| `(4, 1025, 16, 80)` | 0.0930 ms | 0.1243 ms | -25.2% |
| `(16, 256, 16, 64)` | 0.0629 ms | 0.0588 ms | +6.9% |
| `(16, 1025, 16, 64)` | 0.2246 ms | 0.2603 ms | -13.7% |
| `(16, 576, 16, 128)` | 0.1452 ms | 0.1453 ms | -0.1% |
| `(4, 2304, 16, 64)` | 0.2138 ms | 0.2314 ms | -7.6% |
| `(4, 2304, 16, 128)` | 0.3117 ms | 0.3214 ms | -3.0% |

The complete sweep covered `head_dim={64,80,128}`, `seq={256,576,1025,2304}`, and `bs={1,4,16}`. Only `(16,256,16,64)` was measurably faster with cuDNN. All other shapes selected by the proposed policy were approximately neutral or 7-37% slower.

### Real-weight 8B E2E

`OpenGVLab/InternVL2_5-8B`, real BF16 weights, 64 distinct 448x448 image requests per trial, 16 output tokens, three trials per concurrency. `dense cuDNN` temporarily restores the old dense `SingletonCache` behavior so the intended cuDNN branch is actually exercised.

| concurrency | FA3 req/s | current dynamic req/s | dense cuDNN req/s | dense throughput vs FA3 | dense mean E2E vs FA3 |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | 4.5632 | 4.6063 | 4.6173 | +1.19% | -1.17% |
| 4 | 7.8217 | 7.7919 | 7.7708 | -0.65% | +0.65% |
| 16 | 9.2480 | 9.1878 | 9.1999 | -0.52% | +0.53% |

The current dynamic path is an FA3 fallback control and itself varies by roughly 1%, so the concurrency-1 difference is benchmark noise. There is no stable E2E gain at higher concurrency.

### Numerical and quality checks

- Maximum BF16 attention output difference across the microbenchmark shapes: approximately `9.77e-4` to `3.91e-3`.
- Exact greedy output parity between FA3 and forced dense cuDNN: `58.59%`, `64.84%`, and `61.72%` at concurrency 1, 4, and 16.
- Full 900-example MMMU validation, two runs per backend:

| backend | run 1 | run 2 | mean |
| --- | ---: | ---: | ---: |
| FA3 | 31.997% | 31.777% | 31.887% |
| dense cuDNN | 31.000% | 31.550% | 31.275% |

The `-0.61` percentage-point mean difference is not sufficient to claim a quality regression because the concurrent evaluation itself varies, but it also does not establish that changing the default is quality-neutral.

### Closure rationale

Closing this PR because the original optimization premise no longer holds on current main:

1. The original InternVL E2E workload no longer dispatches to cuDNN.
2. The previous apparent gain was largely explained by a repeated `max_seqlen.item()` synchronization that main has since removed.
3. Broader microbenchmarks show cuDNN is slower for nearly all intended policy shapes.
4. Real-weight 8B E2E shows no stable throughput or latency improvement.
5. The backend change is not bitwise transparent, while its quality impact is not sufficiently established to justify changing the Hopper default.

The explicit cuDNN backend or a dynamic policy can be revisited in a new PR if a current model/hardware workload demonstrates a reproducible gain over the current FA3 path.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #30347585778](https://github.com/sgl-project/sglang/actions/runs/30347585778)
Latest PR Test (Extra): :x: [Run #30347585841](https://github.com/sgl-project/sglang/actions/runs/30347585841)

### CODE REVIEW

**Comment by mickqian** [comment]
Closing based on the follow-up validation now added to the PR description.

After merging current main, the original InternVL workload no longer selects cuDNN because its tensor `cu_seqlens` makes the dynamic backend fall back to FA3. Main also precomputes `max_seqlen`, removing the repeated host synchronization that explained much of the earlier apparent gain. The expanded H100 sweep and real-weight InternVL2.5-8B E2E benchmark show no stable performance benefit, while the backend is not bitwise transparent and the quality impact is not sufficiently established.

Given that evidence, changing Hopper's default away from FA3 is not justified. A new PR can revisit an explicit or dynamic cuDNN backend if a current real-model workload demonstrates a reproducible gain over the current FA3 implementation.
